### PR TITLE
Make elapse time optional

### PR DESCRIPTION
**Pull request by xi-liu-ds**
This PR makes elapse time optional with `--outelapsetime` flag set

### CODE REVIEW

File: Cilsil/Program.cs [hunk 1]
```diff
@@ -127,9 +129,12 @@ private static void Main(string[] args)
...
         /// <param name="printprocs">The printprocs.</param>
+        /// <param name="outelapsetime">The elapse time output path (used for visualizing the elapse 
+        /// time per method).</param>
         /// <returns></returns>
         public static (Cfg, TypeEnvironment) ExecuteTranslation(string[] paths,
-                                                                string printprocs = null)
+                                                                string printprocs = null,
+                                                                string outelapsetime = null)
         {
             var watch = System.Diagnostics.Stopwatch.StartNew();
```
Comment (xinshiMSFT) [inline Cilsil/Program.cs]: we don't want to start this resource at all when the flag is off

File: Cilsil/Program.cs [hunk 2]
```diff
@@ -154,13 +159,17 @@ private static void Main(string[] args)
...
-
             Log.PrintAllUnknownInstruction();
             Log.WriteLine();
             Log.PrintCoverageStats(result.GetResult<CfgParserResult>().Methods);
-            Log.WriteLine();
-            Log.PrintProcessTime(watch.ElapsedMilliseconds);
+            
+            watch.Stop();
+
+            if (!string.IsNullOrWhiteSpace(outelapsetime))
```
Comment (xinshiMSFT) [inline Cilsil/Program.cs]: use Microsoft.FeatureManager which is directly targeting this scenario rather than directly comparing string using if

**Review by xinshiMSFT — CHANGES_REQUESTED** [review]

File: Cilsil/Program.cs (hunk 2 above)
Comment (xi-liu-ds) [reply Cilsil/Program.cs]: Microsoft.FeatureManager is for more complicated case. I update with one debug flag

File: Cilsil/Program.cs (hunk 1 above)
Comment (xi-liu-ds) [reply Cilsil/Program.cs]: Done

File: Cilsil/Program.cs [hunk 3]
```diff
@@ -92,26 +95,28 @@ private static void Main(string[] args)
...
         {
+            Log.Debug = debug;
             (var cfg, var tenv) = ExecuteTranslation(paths, printprocs);
 
             File.WriteAllText(cfgtxt ?? "./cfg.txt", cfg.ToString());
             cfg.WriteToFile(outcfg);
             tenv.WriteToFile(outtenv);
 
-            if (!string.IsNullOrWhiteSpace(outelapsetime))
+            if (debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Program.cs]: nit: just debug

File: Cilsil/Program.cs [hunk 4]
```diff
@@ -131,7 +136,11 @@ private static void Main(string[] args)
         public static (Cfg, TypeEnvironment) ExecuteTranslation(string[] paths,
                                                                 string printprocs = null)
         {
-            var watch = System.Diagnostics.Stopwatch.StartNew();
+            System.Diagnostics.Stopwatch watch = null;
+            if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Program.cs]: same

File: Cilsil/Program.cs [hunk 5]
```diff
@@ -154,13 +163,17 @@ private static void Main(string[] args)
...
 
-            watch.Stop();
-
             Log.PrintAllUnknownInstruction();
             Log.WriteLine();
             Log.PrintCoverageStats(result.GetResult<CfgParserResult>().Methods);
-            Log.WriteLine();
-            Log.PrintProcessTime(watch.ElapsedMilliseconds);
+            
+            if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Program.cs]: same

File: Cilsil/Services/CfgParserService.cs [hunk 6]
```diff
@@ -23,13 +23,17 @@ internal class CfgParserService : IService
 
         public IEnumerable<TypeDefinition> Types { get; private set; }
 
+        public string OutElapseTime { get; private set; }
```
Comment (xinshiMSFT) [inline Cilsil/Services/CfgParserService.cs]: use timespan instead of string

File: Cilsil/Services/CfgParserService.cs [hunk 7]
```diff
@@ -52,13 +56,18 @@ public ServiceExecutionResult Execute()
             Cfg = new Cfg();
             foreach (var method in Methods)
             {
-                var watch = System.Diagnostics.Stopwatch.StartNew();
+                System.Diagnostics.Stopwatch watch = null;
+                if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Services/CfgParserService.cs]: same

File: Cilsil/Services/CfgParserService.cs [hunk 8]
```diff
@@ -207,6 +216,12 @@ private bool ComputeMethodCfg(MethodDefinition method)
             state.PushInstruction(instruction);
             do
             {
+                System.Diagnostics.Stopwatch watch = null;
+                if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Services/CfgParserService.cs]: same

File: Cilsil/Services/CfgParserService.cs [hunk 9]
```diff
@@ -245,6 +260,12 @@ private bool ComputeMethodCfg(MethodDefinition method)
                     break;
                 }
 
+                if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Services/CfgParserService.cs]: same

File: Cilsil/Utils/ProgramState.cs [hunk 10]
```diff
@@ -418,7 +418,10 @@ public void PushRetExpr()
                 }
                 CurrentLocation = newLocation;
             }
-            ParsedInstructions.Add(snapshot.Instruction);
+            if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Utils/ProgramState.cs]: same

File: Cilsil/Utils/ProgramState.cs [hunk 11]
```diff
@@ -430,11 +433,18 @@ public void PushRetExpr()
         /// <returns>String representing the debug information.</returns>
         public string GetStateDebugInformation(object invalidObject)
         {
-            return $"Invalid value {invalidObject?.ToString()}\n" +
-                    "====State information====\n" +
-                    ProcDesc.ToString() + "\n" +
-                    "====Parsed Instructions====\n" +
-                    string.Join(",", ParsedInstructions);
+            if (Log.Debug is true)
```
Comment (xinshiMSFT) [inline Cilsil/Utils/ProgramState.cs]: same

File: Cilsil/Utils/ProgramState.cs [hunk 12]
```diff
@@ -430,11 +433,18 @@ public void PushRetExpr()
...
+            { 
+                return $"Invalid value {invalidObject?.ToString()}\n" +
+                        "====State information====\n" +
+                        ProcDesc.ToString() + "\n" +
+                        "====Parsed Instructions====\n" +
+                        string.Join(",", ParsedInstructions);
+            }
+            else
+            {
+                return "";
```
Comment (xinshiMSFT) [inline Cilsil/Utils/ProgramState.cs]: nit: string.empty

Comment (xi-liu-ds) [reply Cilsil/Utils/ProgramState.cs]: Done

File: Cilsil/Utils/ProgramState.cs (hunk 11 above)
Comment (xi-liu-ds) [reply Cilsil/Utils/ProgramState.cs]: Done

File: Cilsil/Program.cs (hunk 3 above)
Comment (xi-liu-ds) [reply Cilsil/Program.cs]: Done

File: Cilsil/Program.cs (hunk 4 above)
Comment (xi-liu-ds) [reply Cilsil/Program.cs]: Done

File: Cilsil/Program.cs (hunk 5 above)
Comment (xi-liu-ds) [reply Cilsil/Program.cs]: Done

File: Cilsil/Services/CfgParserService.cs (hunk 7 above)
Comment (xi-liu-ds) [reply Cilsil/Services/CfgParserService.cs]: Done

File: Cilsil/Services/CfgParserService.cs (hunk 8 above)
Comment (xi-liu-ds) [reply Cilsil/Services/CfgParserService.cs]: Done

File: Cilsil/Services/CfgParserService.cs (hunk 9 above)
Comment (xi-liu-ds) [reply Cilsil/Services/CfgParserService.cs]: Done

File: Cilsil/Utils/ProgramState.cs (hunk 10 above)
Comment (xi-liu-ds) [reply Cilsil/Utils/ProgramState.cs]: Done

File: Cilsil/Services/CfgParserService.cs (hunk 6 above)
Comment (xi-liu-ds) [reply Cilsil/Services/CfgParserService.cs]: Removed due to legacy code

**Review by xinshiMSFT — APPROVED** [review]

**Pull request merged**